### PR TITLE
No more item spam attacks

### PR DESCRIPTION
**Pull request by Allfd**
This is to address the last remaining death blenders.

🆑 Alffd
add: Humans are knocked down for two seconds when hit by an item blender object that is strong enough to hurt them.  This will have them down long enough for the blender to subside.
add: Borgs when hit by item blender objects, will have a 80% chance of taking no damage, and a 20% chance of taking damage from the hit.
add: Xenos when hit by item blender objects, will have a 90% chance of dodging and a 10% chance of taking a hit and getting knocked down for 1 second.
/ 🆑 

In my last test, it was possible to survive a swarm of 45 glass shards while only taking 20 brute in the case of humans and xenos.

Borgs took 90 brute.

As it stands now, the mob would take 450 brute.

Bask in the reflected glory of the Nanotrasen materials safety test chamber!

image

### CODE REVIEW

File: code/modules/mob/living/carbon/alien/humanoid/humanoid_defense.dm [hunk 1]
```diff
@@ -1,3 +1,18 @@
+/mob/living/carbon/alien/humanoid/hitby(atom/movable/AM, skipcatch = 0, hitpush = 0, blocked = 0)
+	hitpush = 0
+	var/obj/item/I
+	if(istype(AM, /obj/item))
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/alien/humanoid/humanoid_defense.dm]: This should be an `!istype()` check with an inner body of `return ..()`

File: code/modules/mob/living/carbon/alien/humanoid/humanoid_defense.dm [hunk 2]
```diff
@@ -1,3 +1,18 @@
...
+		I = AM
+	if(I)
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/alien/humanoid/humanoid_defense.dm]: This should be better composed with `&&` so it's not a mountain of tabs and if statements

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 3]
```diff
@@ -316,6 +316,10 @@ emp_act
 					visible_message("<span class='danger'>[I] embeds itself in [src]'s [L.name]!</span>","<span class='userdanger'>[I] embeds itself in your [L.name]!</span>")
 					hitpush = 0
 					skipcatch = 1 //can't catch the now embedded item
+		if(I.throwforce)
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Compose these conditionals together

File: code/modules/mob/living/silicon/robot/robot_defense.dm [hunk 4]
```diff
@@ -57,3 +59,15 @@
 				step_away(src, user, 15)
 				sleep(3)
 				step_away(src, user, 15)
+
+/mob/living/silicon/robot/hitby(atom/movable/AM, skipcatch = 0, hitpush = 1, blocked = 0)
+	var/obj/item/I
+	if(istype(AM, /obj/item))
+		I = AM
+	if(I)
```
Comment (Crazylemon64) [inline code/modules/mob/living/silicon/robot/robot_defense.dm]: Compose these conditionals together. Additionally, make the `istype` negated and do a parent call inside, as above.

**Comment by pinatacolada** [comment]
Ai cores also suffer from blenders

**Comment by Allfd** [comment]
@pinatacolada 

> Ai cores also suffer from blenders

They do, but this is a balance PR, most people don't like the death blenders, but think that mass items should cause harm.  Its also exceptionally hard to tweak as the atmospheric systems is super dynamic.  Which makes for really cool effects and uses, but also means that player choice has a massive impact on just how deadly it will be.

TLDR; there is going to be a fight over numbers and effects on this, so I kept it small to begin with.

**Comment by Crazylemon64** [comment]
There's a lot of other things than atmos that don't set the thrower - such as gibbing, pneumatic cannons, disposals launching, sorium, dark matter...

It'd be better to `#define` a unique identifier for atmos throwing that would then be compared to for this, rather than simply cause a weaken on any non-mob throwing

**Comment by Allfd** [comment]
@Crazylemon64 
I had a version that did that, but it ended up touching a lot of other things in a bunch of different places and required or a new variable for all throwing, or a mob to act as atmos

**Comment by Crazylemon64** [comment]
I'm afraid I am unable to accept this PR in this state if it's going to leave side-effects like that.

File: code/modules/mob/living/carbon/alien/humanoid/humanoid_defense.dm (hunk 1 above)
Comment (Allfd) [reply code/modules/mob/living/carbon/alien/humanoid/humanoid_defense.dm]: Doing so will break things, as some of this behavior is dependent on calls up the path.

Removing the notification though, as that is spamming, and can conflict with later notices.

File: code/game/objects/items/weapons/dice.dm [hunk 5]
```diff
@@ -111,7 +111,7 @@
 /obj/item/dice/attack_self(mob/user as mob)
 	diceroll(user)
 
-/obj/item/dice/throw_at(atom/target, range, speed, mob/thrower, spin=1, diagonals_first = 0, datum/callback/callback)
+/obj/item/dice/throw_at(atom/target, range, speed, mob/thrower, spin=1, diagonals_first = 0, datum/callback/callback, var/atmos = FALSE)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/dice.dm]: Please avoid adding `var/` prefixes to arguments.

File: code/game/objects/items.dm [hunk 6]
```diff
@@ -500,10 +500,10 @@ var/global/image/fire_overlay = image("icon" = 'icons/goonstation/effects/fire.d
 			itempush = 0 // too light to push anything
 		return A.hitby(src, 0, itempush)
 
-/obj/item/throw_at(atom/target, range, speed, mob/thrower, spin=1, diagonals_first = 0, datum/callback/callback)
+/obj/item/throw_at(atom/target, range, speed, mob/thrower, spin=1, diagonals_first = 0, datum/callback/callback, var/atmos = FALSE)
```
Comment (Crazylemon64) [inline code/game/objects/items.dm]: Please avoid adding `var/` prefixes to arguments.

File: code/game/atoms_movable.dm [hunk 7]
```diff
@@ -221,10 +221,15 @@
 		step(src, AM.dir)
 	..()
 
-/atom/movable/proc/throw_at(atom/target, range, speed, mob/thrower, spin = TRUE, diagonals_first = FALSE, var/datum/callback/callback)
+/atom/movable/proc/throw_at(atom/target, range, speed, mob/thrower, spin = TRUE, diagonals_first = FALSE, var/datum/callback/callback, var/atmos = FALSE)
```
Comment (Crazylemon64) [inline code/game/atoms_movable.dm]: Please avoid adding `var/` prefixes to arguments.

**Comment by Crazylemon64** [comment]
This PR requires an update

**Comment by tigercat2000** [comment]
Please make the changes Crazy requested, or this PR will be closed in one week (other maints may correct me if the stale timer is different).

**Comment by Allfd** [comment]
My understanding of this, is that it takes mob/

I may not be a expert on this, but doesn't that get checked for logging?  How do you jump to "Atmos"?

**Comment by Crazylemon64** [comment]
@Allfd BYOND uses duck typing

meaning that if it never actually asks it to quack or float, it can't tell if you gave it a duck or not

So if you feed it "Atmos", the type guards will prevent any code from running that would choke, while you would also be able to compare to it in order to run special atmos throwing logic.

**Comment by variableundefined** [comment]
@Allfd In order to clean up the issue tracker I am closing down PR that requires an update and is not yet merge-ready. You are free to re-open this PR when it is ready. 

The only thing we are objecting to is the handling of the PR in the code. No token will be charged when this PR is re-opened and it will go through a review by the maintainers.
